// @fleet-sdk/core/dist/esm/serializer/sigma/sigmaTypes.js
import { ensureBigInt, isDefined } from "@fleet-sdk/common";
import { SigmaTypeCode } from "./sigmaTypeCode";
export function SByte(value) {
    return _createPrimitiveType(SigmaTypeCode.Byte, value);
}
export function SBool(value) {
    return _createPrimitiveType(SigmaTypeCode.Boolean, value);
}
export function SShort(value) {
    return _createPrimitiveType(SigmaTypeCode.Short, value);
}
export function SInt(value) {
    return _createPrimitiveType(SigmaTypeCode.Int, value);
}
export function SLong(value) {
    return _createPrimitiveType(SigmaTypeCode.Long, isDefined(value) ? ensureBigInt(value) : undefined);
}
export function SBigInt(value) {
    return _createPrimitiveType(SigmaTypeCode.BigInt, isDefined(value) ? ensureBigInt(value) : undefined);
}
export function SUnit() {
    return _createPrimitiveType(SigmaTypeCode.Unit, null);
}
export function SGroupElement(value) {
    return _createPrimitiveType(SigmaTypeCode.GroupElement, value);
}
export function SSigmaProp(value) {
    return _createPrimitiveType(SigmaTypeCode.SigmaProp, value);
}
function _createPrimitiveType(type, value) {
    if (value !== undefined) {
        return { type, value };
    }
    else {
        return type;
    }
}
export function SColl(type, elements) {
    return {
        type: SigmaTypeCode.Coll,
        elementsType: type(),
        value: elements
    };
}
