// @fleet-sdk/core/dist/esm/builder/outputBuilder.js
import { _0n } from "@fleet-sdk/common";
import { areRegistersDenselyPacked, ensureBigInt, first, isEmpty, isHex, isUndefined, removeUndefined } from "@fleet-sdk/common";
import { stringToBytes } from "@scure/base";
import { InvalidRegistersPacking } from "../errors/invalidRegistersPacking";
import { UndefinedCreationHeight } from "../errors/undefinedCreationHeight";
import { UndefinedMintingContext } from "../errors/undefinedMintingContext";
import { ErgoAddress } from "../models";
import { TokensCollection } from "../models/collections/tokensCollection";
import { SConstant } from "../serializer/sigma/constantSerializer";
import { SByte, SColl } from "../serializer/sigma/sigmaTypes";
export const SAFE_MIN_BOX_VALUE = BigInt(1000000);
export class OutputBuilder {
    _value;
    _address;
    _tokens;
    _creationHeight;
    _registers;
    _minting;
    constructor(value, recipient, creationHeight) {
        this._value = ensureBigInt(value);
        this._creationHeight = creationHeight;
        this._tokens = new TokensCollection();
        this._registers = {};
        if (this._value <= _0n) {
            throw new Error("An UTxO cannot be created without a minimum required amount.");
        }
        if (typeof recipient === "string") {
            this._address = isHex(recipient)
                ? ErgoAddress.fromErgoTree(recipient)
                : ErgoAddress.fromBase58(recipient);
        }
        else {
            this._address = recipient;
        }
    }
    get value() {
        return this._value;
    }
    get address() {
        return this._address;
    }
    get ergoTree() {
        return this._address.ergoTree;
    }
    get creationHeight() {
        return this._creationHeight;
    }
    get tokens() {
        return this._tokens;
    }
    get additionalRegisters() {
        return this._registers;
    }
    get minting() {
        return this._minting;
    }
    addTokens(tokens, options) {
        if (tokens instanceof TokensCollection) {
            this._tokens.add(tokens.toArray(), options);
        }
        else {
            this._tokens.add(tokens, options);
        }
        return this;
    }
    mintToken(token) {
        this._minting = { ...token, amount: ensureBigInt(token.amount) };
        return this;
    }
    setCreationHeight(height, options) {
        if (isUndefined(options) ||
            options.replace === true ||
            (options.replace === false && isUndefined(this._creationHeight))) {
            this._creationHeight = height;
        }
        return this;
    }
    setAdditionalRegisters(registers) {
        this._registers = removeUndefined(registers);
        if (!areRegistersDenselyPacked(registers)) {
            throw new InvalidRegistersPacking();
        }
        return this;
    }
    eject(ejector) {
        ejector({ tokens: this._tokens });
        return this;
    }
    build(transactionInputs) {
        let tokens = this.tokens.toArray();
        if (this.minting) {
            if (isEmpty(transactionInputs)) {
                throw new UndefinedMintingContext();
            }
            if (isEmpty(this.additionalRegisters)) {
                this.setAdditionalRegisters({
                    R4: SConstant(SColl(SByte, stringToBytes("utf8", this.minting.name || ""))),
                    R5: SConstant(SColl(SByte, stringToBytes("utf8", this.minting.description || ""))),
                    R6: SConstant(SColl(SByte, stringToBytes("utf8", this.minting.decimals?.toString() || "0")))
                });
            }
            tokens = [
                {
                    tokenId: first(transactionInputs).boxId,
                    amount: this.minting.amount
                },
                ...tokens
            ];
        }
        if (isUndefined(this.creationHeight)) {
            throw new UndefinedCreationHeight();
        }
        return {
            value: this.value.toString(),
            ergoTree: this.ergoTree,
            creationHeight: this.creationHeight,
            assets: tokens.map((token) => {
                return {
                    tokenId: token.tokenId,
                    amount: token.amount.toString()
                };
            }),
            additionalRegisters: this.additionalRegisters
        };
    }
}
