// @fleet-sdk/core/dist/esm/builder/transactionBuilderSettings.js
import { MAX_TOKENS_PER_BOX } from "../models";
export class TransactionBuilderSettings {
    _maxDistinctTokensPerChangeBox;
    _allowTokenBurn;
    _allowTokenPluginFromPlugins;
    constructor() {
        this._maxDistinctTokensPerChangeBox = MAX_TOKENS_PER_BOX;
        this._allowTokenBurn = false;
        this._allowTokenPluginFromPlugins = false;
    }
    get maxTokensPerChangeBox() {
        return this._maxDistinctTokensPerChangeBox;
    }
    get canBurnTokens() {
        return this._allowTokenBurn;
    }
    get canBurnTokensFromPlugins() {
        return this.canBurnTokens || this._allowTokenPluginFromPlugins;
    }
    /**
     * Define max number of distinct tokens per change box
     */
    setMaxTokensPerChangeBox(max) {
        this._maxDistinctTokensPerChangeBox = max;
        return this;
    }
    /**
     * Allows or denies token burning from all contexts
     */
    allowTokenBurning(allow) {
        this._allowTokenBurn = allow;
        return this;
    }
    /**
     * Allows or denies token burning **only** from plugins context.
     */
    allowTokenBurningFromPlugins(allow) {
        this._allowTokenPluginFromPlugins = allow;
        return this;
    }
}
