// @fleet-sdk/core/dist/cjs/builder/outputBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutputBuilder = exports.SAFE_MIN_BOX_VALUE = void 0;
const common_1 = require("@fleet-sdk/common");
const common_2 = require("@fleet-sdk/common");
const base_1 = require("@scure/base");
const invalidRegistersPacking_1 = require("../errors/invalidRegistersPacking");
const undefinedCreationHeight_1 = require("../errors/undefinedCreationHeight");
const undefinedMintingContext_1 = require("../errors/undefinedMintingContext");
const models_1 = require("../models");
const tokensCollection_1 = require("../models/collections/tokensCollection");
const constantSerializer_1 = require("../serializer/sigma/constantSerializer");
const sigmaTypes_1 = require("../serializer/sigma/sigmaTypes");
exports.SAFE_MIN_BOX_VALUE = BigInt(1000000);
class OutputBuilder {
    _value;
    _address;
    _tokens;
    _creationHeight;
    _registers;
    _minting;
    constructor(value, recipient, creationHeight) {
        this._value = (0, common_2.ensureBigInt)(value);
        this._creationHeight = creationHeight;
        this._tokens = new tokensCollection_1.TokensCollection();
        this._registers = {};
        if (this._value <= common_1._0n) {
            throw new Error("An UTxO cannot be created without a minimum required amount.");
        }
        if (typeof recipient === "string") {
            this._address = (0, common_2.isHex)(recipient)
                ? models_1.ErgoAddress.fromErgoTree(recipient)
                : models_1.ErgoAddress.fromBase58(recipient);
        }
        else {
            this._address = recipient;
        }
    }
    get value() {
        return this._value;
    }
    get address() {
        return this._address;
    }
    get ergoTree() {
        return this._address.ergoTree;
    }
    get creationHeight() {
        return this._creationHeight;
    }
    get tokens() {
        return this._tokens;
    }
    get additionalRegisters() {
        return this._registers;
    }
    get minting() {
        return this._minting;
    }
    addTokens(tokens, options) {
        if (tokens instanceof tokensCollection_1.TokensCollection) {
            this._tokens.add(tokens.toArray(), options);
        }
        else {
            this._tokens.add(tokens, options);
        }
        return this;
    }
    mintToken(token) {
        this._minting = { ...token, amount: (0, common_2.ensureBigInt)(token.amount) };
        return this;
    }
    setCreationHeight(height, options) {
        if ((0, common_2.isUndefined)(options) ||
            options.replace === true ||
            (options.replace === false && (0, common_2.isUndefined)(this._creationHeight))) {
            this._creationHeight = height;
        }
        return this;
    }
    setAdditionalRegisters(registers) {
        this._registers = (0, common_2.removeUndefined)(registers);
        if (!(0, common_2.areRegistersDenselyPacked)(registers)) {
            throw new invalidRegistersPacking_1.InvalidRegistersPacking();
        }
        return this;
    }
    eject(ejector) {
        ejector({ tokens: this._tokens });
        return this;
    }
    build(transactionInputs) {
        let tokens = this.tokens.toArray();
        if (this.minting) {
            if ((0, common_2.isEmpty)(transactionInputs)) {
                throw new undefinedMintingContext_1.UndefinedMintingContext();
            }
            if ((0, common_2.isEmpty)(this.additionalRegisters)) {
                this.setAdditionalRegisters({
                    R4: (0, constantSerializer_1.SConstant)((0, sigmaTypes_1.SColl)(sigmaTypes_1.SByte, (0, base_1.stringToBytes)("utf8", this.minting.name || ""))),
                    R5: (0, constantSerializer_1.SConstant)((0, sigmaTypes_1.SColl)(sigmaTypes_1.SByte, (0, base_1.stringToBytes)("utf8", this.minting.description || ""))),
                    R6: (0, constantSerializer_1.SConstant)((0, sigmaTypes_1.SColl)(sigmaTypes_1.SByte, (0, base_1.stringToBytes)("utf8", this.minting.decimals?.toString() || "0")))
                });
            }
            tokens = [
                {
                    tokenId: (0, common_2.first)(transactionInputs).boxId,
                    amount: this.minting.amount
                },
                ...tokens
            ];
        }
        if ((0, common_2.isUndefined)(this.creationHeight)) {
            throw new undefinedCreationHeight_1.UndefinedCreationHeight();
        }
        return {
            value: this.value.toString(),
            ergoTree: this.ergoTree,
            creationHeight: this.creationHeight,
            assets: tokens.map((token) => {
                return {
                    tokenId: token.tokenId,
                    amount: token.amount.toString()
                };
            }),
            additionalRegisters: this.additionalRegisters
        };
    }
}
exports.OutputBuilder = OutputBuilder;
