// @fleet-sdk/core/dist/cjs/builder/transactionBuilderSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionBuilderSettings = void 0;
const models_1 = require("../models");
class TransactionBuilderSettings {
    _maxDistinctTokensPerChangeBox;
    _allowTokenBurn;
    _allowTokenPluginFromPlugins;
    constructor() {
        this._maxDistinctTokensPerChangeBox = models_1.MAX_TOKENS_PER_BOX;
        this._allowTokenBurn = false;
        this._allowTokenPluginFromPlugins = false;
    }
    get maxTokensPerChangeBox() {
        return this._maxDistinctTokensPerChangeBox;
    }
    get canBurnTokens() {
        return this._allowTokenBurn;
    }
    get canBurnTokensFromPlugins() {
        return this.canBurnTokens || this._allowTokenPluginFromPlugins;
    }
    /**
     * Define max number of distinct tokens per change box
     */
    setMaxTokensPerChangeBox(max) {
        this._maxDistinctTokensPerChangeBox = max;
        return this;
    }
    /**
     * Allows or denies token burning from all contexts
     */
    allowTokenBurning(allow) {
        this._allowTokenBurn = allow;
        return this;
    }
    /**
     * Allows or denies token burning **only** from plugins context.
     */
    allowTokenBurningFromPlugins(allow) {
        this._allowTokenPluginFromPlugins = allow;
        return this;
    }
}
exports.TransactionBuilderSettings = TransactionBuilderSettings;
