// @fleet-sdk/core/dist/esm/builder/selector/boxSelector.js
import { first, isUndefined } from "@fleet-sdk/common";
import { _0n, ensureBigInt, hasDuplicatesBy, isDefined, isEmpty, orderBy, some, sumBy, utxoSum } from "@fleet-sdk/common";
import { DuplicateInputSelectionError } from "../../errors/duplicateInputSelectionError";
import { InsufficientInputs } from "../../errors/insufficientInputs";
import { AccumulativeSelectionStrategy } from "./strategies/accumulativeSelectionStrategy";
import { CustomSelectionStrategy } from "./strategies/customSelectionStrategy";
export class BoxSelector {
    _inputs;
    _strategy;
    _ensureFilterPredicate;
    _inputsSortSelector;
    _inputsSortDir;
    _ensureInclusionBoxIds;
    constructor(inputs) {
        this._inputs = inputs;
    }
    defineStrategy(strategy) {
        if (this._isISelectionStrategyImplementation(strategy)) {
            this._strategy = strategy;
        }
        else {
            this._strategy = new CustomSelectionStrategy(strategy);
        }
        return this;
    }
    select(target) {
        if (!this._strategy) {
            this._strategy = new AccumulativeSelectionStrategy();
        }
        const remaining = this._deepCloneTarget(target);
        let unselected = [...this._inputs];
        let selected = [];
        const predicate = this._ensureFilterPredicate;
        const inclusion = this._ensureInclusionBoxIds;
        if (isDefined(predicate)) {
            if (isDefined(inclusion)) {
                selected = unselected.filter((box) => predicate(box) || inclusion.has(box.boxId));
            }
            else {
                selected = unselected.filter(predicate);
            }
        }
        else if (isDefined(inclusion)) {
            selected = unselected.filter((box) => inclusion.has(box.boxId));
        }
        if (isDefined(selected)) {
            unselected = unselected.filter((box) => !selected.some((sel) => sel.boxId === box.boxId));
            if (isDefined(remaining.nanoErgs)) {
                remaining.nanoErgs -= sumBy(selected, (input) => input.value);
            }
            if (isDefined(remaining.tokens) && selected.some((input) => !isEmpty(input.assets))) {
                for (const tokenTarget of remaining.tokens) {
                    if (tokenTarget.amount) {
                        tokenTarget.amount -= utxoSum(selected, tokenTarget.tokenId);
                    }
                }
            }
        }
        unselected = this._sort(unselected);
        selected = selected.concat(this._strategy.select(unselected, remaining));
        if (hasDuplicatesBy(selected, (item) => item.boxId)) {
            throw new DuplicateInputSelectionError();
        }
        const unreached = this._getUnreachedTargets(selected, target);
        if (unreached.nanoErgs || some(unreached.tokens)) {
            throw new InsufficientInputs(unreached);
        }
        return selected;
    }
    _deepCloneTarget(target) {
        return {
            nanoErgs: target.nanoErgs,
            tokens: isDefined(target.tokens)
                ? target.tokens.map((t) => ({ tokenId: t.tokenId, amount: t.amount }))
                : undefined
        };
    }
    _getUnreachedTargets(inputs, target) {
        const unreached = { nanoErgs: undefined, tokens: undefined };
        const selectedNanoergs = sumBy(inputs, (input) => input.value);
        if (target.nanoErgs && target.nanoErgs > selectedNanoergs) {
            unreached.nanoErgs = target.nanoErgs - selectedNanoergs;
        }
        if (isEmpty(target.tokens)) {
            return unreached;
        }
        for (const tokenTarget of target.tokens) {
            const totalSelected = utxoSum(inputs, tokenTarget.tokenId);
            if (isDefined(tokenTarget.amount) && tokenTarget.amount > totalSelected) {
                if (tokenTarget.tokenId === first(inputs).boxId) {
                    continue;
                }
                if (isUndefined(unreached.tokens)) {
                    unreached.tokens = [];
                }
                unreached.tokens.push({
                    tokenId: tokenTarget.tokenId,
                    amount: tokenTarget.amount - totalSelected
                });
            }
        }
        return unreached;
    }
    _sort(inputs) {
        if (!this._inputsSortSelector) {
            return orderBy(inputs, (input) => input.creationHeight, "asc");
        }
        return orderBy(inputs, this._inputsSortSelector, this._inputsSortDir || "asc");
    }
    ensureInclusion(predicateOrBoxIds) {
        if (typeof predicateOrBoxIds === "function") {
            this._ensureFilterPredicate = predicateOrBoxIds;
        }
        else {
            if (isUndefined(this._ensureInclusionBoxIds)) {
                this._ensureInclusionBoxIds = new Set();
            }
            if (Array.isArray(predicateOrBoxIds)) {
                for (const boxId of predicateOrBoxIds) {
                    this._ensureInclusionBoxIds.add(boxId);
                }
            }
            else {
                this._ensureInclusionBoxIds.add(predicateOrBoxIds);
            }
        }
        return this;
    }
    orderBy(selector, direction) {
        this._inputsSortSelector = selector;
        this._inputsSortDir = direction;
        return this;
    }
    _isISelectionStrategyImplementation(obj) {
        if (obj.select) {
            return true;
        }
        return false;
    }
    static buildTargetFrom(boxes) {
        const tokens = {};
        let nanoErgs = _0n;
        for (const box of boxes) {
            nanoErgs += ensureBigInt(box.value);
            for (const token of box.assets) {
                tokens[token.tokenId] = (tokens[token.tokenId] || _0n) + ensureBigInt(token.amount);
            }
        }
        return {
            nanoErgs,
            tokens: Object.keys(tokens).map((tokenId) => ({ tokenId, amount: tokens[tokenId] }))
        };
    }
}
