// @fleet-sdk/core/dist/esm/models/ergoBox.js
import { ensureBigInt } from "@fleet-sdk/common";
import { blake2b } from "@noble/hashes/blake2b";
import { bytesToHex } from "@noble/hashes/utils";
import { serializeBox } from "../serializer/sigma/boxSerializer";
export class ErgoBox {
    boxId;
    value;
    ergoTree;
    creationHeight;
    assets;
    additionalRegisters;
    transactionId;
    index;
    constructor(box) {
        this.boxId = box.boxId;
        this.ergoTree = box.ergoTree;
        this.creationHeight = box.creationHeight;
        this.value = ensureBigInt(box.value);
        this.assets = box.assets.map((asset) => ({
            tokenId: asset.tokenId,
            amount: ensureBigInt(asset.amount)
        }));
        this.additionalRegisters = box.additionalRegisters;
        this.transactionId = box.transactionId;
        this.index = box.index;
    }
    isValid() {
        return ErgoBox.validate(this);
    }
    static validate(box) {
        const bytes = serializeBox(box);
        const hash = bytesToHex(blake2b(bytes, { dkLen: 32 }));
        return box.boxId === hash;
    }
}
