// @fleet-sdk/core/dist/esm/models/ergoAddress.js
import { AddressType, Network } from "@fleet-sdk/common";
import { areEqual, endsWith, first, isDefined, startsWith } from "@fleet-sdk/common";
import { blake2b } from "@noble/hashes/blake2b";
import { bytesToHex, concatBytes, hexToBytes } from "@noble/hashes/utils";
import { base58 } from "@scure/base";
import { InvalidAddress } from "../errors/invalidAddress";
const CHECKSUM_LENGTH = 4;
const BLAKE_256_HASH_LENGTH = 32;
const P2PK_ERGOTREE_PREFIX = hexToBytes("0008cd");
const P2PK_ERGOTREE_LENGTH = 36;
const P2SH_ERGOTREE_SUFFIX = hexToBytes("d40801");
const P2SH_ERGOTREE_PREFIX = hexToBytes("00ea02d193b4cbe4e3010e040004300e18");
const P2SH_ERGOTREE_LENGTH = 44;
const P2SH_HASH_LENGTH = 24;
function _getEncodedNetworkType(addressBytes) {
    return first(addressBytes) & 0xf0;
}
function _getEncodedAddressType(addressBytes) {
    return first(addressBytes) & 0xf;
}
function _ensureBytes(content) {
    if (typeof content === "string") {
        return hexToBytes(content);
    }
    return content;
}
function blake2b256(input) {
    return blake2b(input, { dkLen: BLAKE_256_HASH_LENGTH });
}
function _getErgoTreeType(ergoTree) {
    if (ergoTree.length === P2PK_ERGOTREE_LENGTH && startsWith(ergoTree, P2PK_ERGOTREE_PREFIX)) {
        return AddressType.P2PK;
    }
    else if (ergoTree.length === P2SH_ERGOTREE_LENGTH &&
        startsWith(ergoTree, P2SH_ERGOTREE_PREFIX) &&
        endsWith(ergoTree, P2SH_ERGOTREE_SUFFIX)) {
        return AddressType.P2SH;
    }
    else {
        return AddressType.P2S;
    }
}
/**
 * Ergo address model
 *
 * @example
 * Convert address to ErgoTree hex string
 * ```
 * const address = new Address("9eZ24iqjKywjzAti9RnWWTR3CiNnLJDAcd2MenKodcAfzc8AFTu");
 * console.log(address.ergoTree);
 * ```
 *
 * @example
 * Convert ErgoTree hex string to address string
 * ```
 * const ergoTree = "0008cd026dc059d64a50d0dbf07755c2c4a4e557e3df8afa7141868b3ab200643d437ee7"
 * const address = Address.fromErgoTree(ergoTree).toString();
 * ```
 */
export class ErgoAddress {
    _ergoTree;
    _network;
    _type;
    get network() {
        return this._network;
    }
    /**
     * ErgoTree hex string
     */
    get ergoTree() {
        return bytesToHex(this._ergoTree);
    }
    get type() {
        return this._type;
    }
    /**
     * New instance from ErgoTree bytes
     * @param ergoTree ErgoTree bytes
     */
    constructor(ergoTree, network) {
        this._ergoTree = ergoTree;
        this._network = network;
        this._type = _getErgoTreeType(ergoTree);
    }
    /**
     * Create a new instance from an ErgoTree
     * @param ergoTree ErgoTree hex string
     */
    static fromErgoTree(ergoTree, network) {
        return new ErgoAddress(hexToBytes(ergoTree), network);
    }
    /**
     * Create a new instance from a public key
     * @param publicKey Public key hex string
     */
    static fromPublicKey(publicKey, network) {
        const ergoTree = concatBytes(P2PK_ERGOTREE_PREFIX, _ensureBytes(publicKey));
        return new ErgoAddress(ergoTree, network);
    }
    static fromHash(hash, network) {
        hash = _ensureBytes(hash);
        if (hash.length === BLAKE_256_HASH_LENGTH) {
            hash = hash.subarray(0, P2SH_HASH_LENGTH);
        }
        else if (hash.length != P2SH_HASH_LENGTH) {
            throw Error(`Invalid hash length: ${hash.length}`);
        }
        const ergoTree = concatBytes(P2SH_ERGOTREE_PREFIX, hash, P2SH_ERGOTREE_SUFFIX);
        return new ErgoAddress(ergoTree, network);
    }
    /**
     * Create a new checked instance from an address string
     * @param encodedAddress Address encoded as base58
     */
    static fromBase58(encodedAddress, skipCheck = false) {
        const bytes = base58.decode(encodedAddress);
        if (!skipCheck && !ErgoAddress.validate(bytes)) {
            throw new InvalidAddress(encodedAddress);
        }
        const network = _getEncodedNetworkType(bytes);
        const type = _getEncodedAddressType(bytes);
        const body = bytes.subarray(1, bytes.length - CHECKSUM_LENGTH);
        if (type === AddressType.P2PK) {
            return this.fromPublicKey(body, network);
        }
        else if (type === AddressType.P2SH) {
            return this.fromHash(body, network);
        }
        else {
            return new ErgoAddress(body, network);
        }
    }
    /**
     * Validate an address
     * @param address Address bytes or string
     */
    static validate(address) {
        const bytes = typeof address === "string" ? base58.decode(address) : address;
        if (bytes.length < CHECKSUM_LENGTH) {
            return false;
        }
        const script = bytes.subarray(0, bytes.length - CHECKSUM_LENGTH);
        const checksum = bytes.subarray(bytes.length - CHECKSUM_LENGTH, bytes.length);
        const blakeHash = blake2b256(script);
        const calculatedChecksum = blakeHash.subarray(0, CHECKSUM_LENGTH);
        return areEqual(calculatedChecksum, checksum);
    }
    static getNetworkType(address) {
        return _getEncodedNetworkType(base58.decode(address));
    }
    static getAddressType(address) {
        return _getEncodedAddressType(base58.decode(address));
    }
    getPublicKeys() {
        if (this.type === AddressType.P2PK) {
            return [this._ergoTree.subarray(P2PK_ERGOTREE_PREFIX.length)];
        }
        return [];
    }
    toP2SH(network) {
        if (this.type === AddressType.P2SH) {
            return this.encode();
        }
        const hash = blake2b256(this._ergoTree).subarray(0, P2SH_HASH_LENGTH);
        return this._encode(hash, AddressType.P2SH, network);
    }
    /**
     * Encode address as base58 string
     */
    encode(network) {
        let body;
        if (this.type === AddressType.P2PK) {
            body = first(this.getPublicKeys());
        }
        else if (this.type === AddressType.P2SH) {
            body = this._ergoTree.subarray(P2SH_ERGOTREE_PREFIX.length, P2SH_ERGOTREE_PREFIX.length + P2SH_HASH_LENGTH);
        }
        else {
            body = this._ergoTree;
        }
        return this._encode(body, this.type, network);
    }
    _encode(body, type, network) {
        if (!isDefined(network)) {
            if (isDefined(this.network)) {
                network = this.network;
            }
            else {
                network = Network.Mainnet;
            }
        }
        const head = Uint8Array.from([network + type]);
        body = concatBytes(head, body);
        const checksum = blake2b256(body).subarray(0, CHECKSUM_LENGTH);
        return base58.encode(concatBytes(body, checksum));
    }
    /**
     * Encode address as base58 string
     */
    toString(network) {
        return this.encode(network);
    }
}
