// @fleet-sdk/core/dist/cjs/serializer/sigma/sigmaTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SColl = exports.SSigmaProp = exports.SGroupElement = exports.SUnit = exports.SBigInt = exports.SLong = exports.SInt = exports.SShort = exports.SBool = exports.SByte = void 0;
const common_1 = require("@fleet-sdk/common");
const sigmaTypeCode_1 = require("./sigmaTypeCode");
function SByte(value) {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.Byte, value);
}
exports.SByte = SByte;
function SBool(value) {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.Boolean, value);
}
exports.SBool = SBool;
function SShort(value) {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.Short, value);
}
exports.SShort = SShort;
function SInt(value) {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.Int, value);
}
exports.SInt = SInt;
function SLong(value) {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.Long, (0, common_1.isDefined)(value) ? (0, common_1.ensureBigInt)(value) : undefined);
}
exports.SLong = SLong;
function SBigInt(value) {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.BigInt, (0, common_1.isDefined)(value) ? (0, common_1.ensureBigInt)(value) : undefined);
}
exports.SBigInt = SBigInt;
function SUnit() {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.Unit, null);
}
exports.SUnit = SUnit;
function SGroupElement(value) {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.GroupElement, value);
}
exports.SGroupElement = SGroupElement;
function SSigmaProp(value) {
    return _createPrimitiveType(sigmaTypeCode_1.SigmaTypeCode.SigmaProp, value);
}
exports.SSigmaProp = SSigmaProp;
function _createPrimitiveType(type, value) {
    if (value !== undefined) {
        return { type, value };
    }
    else {
        return type;
    }
}
function SColl(type, elements) {
    return {
        type: sigmaTypeCode_1.SigmaTypeCode.Coll,
        elementsType: type(),
        value: elements
    };
}
exports.SColl = SColl;
