// @fleet-sdk/core/dist/cjs/serializer/sigma/sigmaByteReader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SigmaByteReader = void 0;
const common_1 = require("@fleet-sdk/common");
const utils_1 = require("@noble/hashes/utils");
const vlq_1 = require("../vlq");
const zigZag_1 = require("../zigZag");
class SigmaByteReader {
    _bytes;
    _cursor;
    get isEmpty() {
        return (0, common_1.isEmpty)(this._bytes);
    }
    constructor(bytes) {
        if (typeof bytes === "string") {
            this._bytes = (0, utils_1.hexToBytes)(bytes);
        }
        else {
            this._bytes = bytes;
        }
        this._cursor = 0;
    }
    readBoolean() {
        return this.readByte() === 0x01;
    }
    readByte() {
        return this._bytes[this._cursor++];
    }
    readType() {
        return this.readByte();
    }
    readNumber() {
        return Number((0, zigZag_1.zigZagDecode)((0, vlq_1.vlqDecode)(this)));
    }
    readLong() {
        return (0, zigZag_1.zigZagDecodeBigInt)((0, vlq_1.vlqDecodeBigInt)(this));
    }
}
exports.SigmaByteReader = SigmaByteReader;
