// @fleet-sdk/core/dist/cjs/serializer/sigma/sigmaTypeCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SigmaTypeCode = void 0;
var SigmaTypeCode;
(function (SigmaTypeCode) {
    SigmaTypeCode[SigmaTypeCode["Boolean"] = 1] = "Boolean";
    SigmaTypeCode[SigmaTypeCode["Byte"] = 2] = "Byte";
    SigmaTypeCode[SigmaTypeCode["Short"] = 3] = "Short";
    SigmaTypeCode[SigmaTypeCode["Int"] = 4] = "Int";
    SigmaTypeCode[SigmaTypeCode["Long"] = 5] = "Long";
    SigmaTypeCode[SigmaTypeCode["BigInt"] = 6] = "BigInt";
    SigmaTypeCode[SigmaTypeCode["GroupElement"] = 7] = "GroupElement";
    SigmaTypeCode[SigmaTypeCode["SigmaProp"] = 8] = "SigmaProp";
    SigmaTypeCode[SigmaTypeCode["Coll"] = 12] = "Coll";
    SigmaTypeCode[SigmaTypeCode["NestedColl"] = 24] = "NestedColl";
    SigmaTypeCode[SigmaTypeCode["Option"] = 36] = "Option";
    SigmaTypeCode[SigmaTypeCode["OptionColl"] = 48] = "OptionColl";
    SigmaTypeCode[SigmaTypeCode["Tuple2"] = 60] = "Tuple2";
    SigmaTypeCode[SigmaTypeCode["Tuple3"] = 72] = "Tuple3";
    SigmaTypeCode[SigmaTypeCode["Tuple4"] = 84] = "Tuple4";
    SigmaTypeCode[SigmaTypeCode["TupleN"] = 96] = "TupleN";
    SigmaTypeCode[SigmaTypeCode["Any"] = 97] = "Any";
    SigmaTypeCode[SigmaTypeCode["Unit"] = 98] = "Unit";
    SigmaTypeCode[SigmaTypeCode["Box"] = 99] = "Box";
    SigmaTypeCode[SigmaTypeCode["AvlTree"] = 100] = "AvlTree";
    SigmaTypeCode[SigmaTypeCode["Context"] = 101] = "Context";
    SigmaTypeCode[SigmaTypeCode["Header"] = 104] = "Header";
    SigmaTypeCode[SigmaTypeCode["PreHeader"] = 105] = "PreHeader";
    SigmaTypeCode[SigmaTypeCode["Global"] = 106] = "Global";
})(SigmaTypeCode = exports.SigmaTypeCode || (exports.SigmaTypeCode = {}));
