// @fleet-sdk/core/dist/cjs/serializer/sigma/sigmaByteWriter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SigmaByteWriter = void 0;
const common_1 = require("@fleet-sdk/common");
const utils_1 = require("@noble/hashes/utils");
const vlq_1 = require("../vlq");
const zigZag_1 = require("../zigZag");
class SigmaByteWriter {
    _bytes;
    _cursor;
    get length() {
        return this._cursor;
    }
    constructor(maxLength) {
        this._bytes = new Uint8Array(maxLength);
        this._cursor = 0;
    }
    writeBoolean(value) {
        this.write(value === true ? 0x01 : 0x00);
        return this;
    }
    writeBooleans(elements) {
        for (let i = 0; i < elements.length; i++) {
            this.writeBoolean(elements[i]);
        }
        return this;
    }
    writeNumber(value) {
        this.writeBytes((0, vlq_1.vlqEncode)((0, zigZag_1.zigZagEncode)(value)));
        return this;
    }
    writeLong(value) {
        this.writeBytes((0, vlq_1.vqlEncodeBigInt)((0, zigZag_1.zigZagEncodeBigInt)(value)));
        return this;
    }
    write(byte) {
        this._bytes[this._cursor++] = byte;
        return this;
    }
    writeBytes(bytes) {
        this._bytes.set(bytes, this._cursor);
        this._cursor += bytes.length;
        return this;
    }
    writeHex(hex) {
        if (hex.length % 2) {
            throw new Error("Invalid hex padding");
        }
        for (let i = 0; i < hex.length / 2; i++) {
            const j = i * 2;
            const byte = Number.parseInt(hex.slice(j, j + 2), 16);
            if (Number.isNaN(byte) || byte < 0) {
                throw new Error("Invalid byte sequence");
            }
            this.write(byte);
        }
        return this;
    }
    writeBits(bits) {
        let bitOffset = 0;
        for (let i = 0; i < bits.length; i++) {
            if (bits[i]) {
                this._bytes[this._cursor] |= 1 << bitOffset++;
            }
            else {
                this._bytes[this._cursor] &= ~(1 << bitOffset++);
            }
            if (bitOffset == 8) {
                bitOffset = 0;
                this._cursor++;
            }
        }
        if (bitOffset > 0) {
            this._cursor++;
        }
        return this;
    }
    writeBigInt(number) {
        // todo: take a look at https://coolaj86.com/articles/convert-decimal-to-hex-with-js-bigints/
        // and https://coolaj86.com/articles/convert-hex-to-decimal-with-js-bigints/
        if (number < common_1._0n) {
            throw new Error("Negative BigInt values are not supported Fleet serializer.");
        }
        let hex = number.toString(16);
        if (hex.length % 2) {
            hex = "0" + hex;
        }
        else if (Number.parseInt(hex.substring(0, 1), 16) >= 8) {
            // maximum positive need to prepend 0 otherwise results in negative number
            hex = "00" + hex;
        }
        this.writeBytes((0, vlq_1.vlqEncode)(hex.length / 2));
        this.writeHex(hex);
        return this;
    }
    toHex() {
        return (0, utils_1.bytesToHex)(this._bytes.subarray(0, this._cursor));
    }
    toBytes() {
        return this._bytes.subarray(0, this._cursor);
    }
}
exports.SigmaByteWriter = SigmaByteWriter;
