// @fleet-sdk/core/dist/cjs/models/collections/collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Collection = void 0;
const common_1 = require("@fleet-sdk/common");
class Collection {
    _items;
    constructor() {
        this._items = [];
    }
    _isIndexOutOfBounds(index) {
        return index < 0 || index >= this._items.length;
    }
    [Symbol.iterator]() {
        let counter = 0;
        return {
            next: () => {
                return {
                    done: counter >= this.length,
                    value: this._items[counter++]
                };
            }
        };
    }
    get length() {
        return this._items.length;
    }
    get isEmpty() {
        return this.length === 0;
    }
    at(index) {
        if (this._isIndexOutOfBounds(index)) {
            throw new RangeError(`Index '${index}' is out of range.`);
        }
        return this._items[index];
    }
    add(items, options) {
        return this._addOneOrMore(items, options);
    }
    _addOne(item, options) {
        if ((0, common_1.isDefined)(options) && (0, common_1.isDefined)(options.index)) {
            if (options.index === this.length) {
                this._items.push(this._map(item));
                return this.length;
            }
            if (this._isIndexOutOfBounds(options.index)) {
                throw new RangeError(`Index '${options.index}' is out of range.`);
            }
            this._items.splice(options.index, 0, this._map(item));
            return this.length;
        }
        this._items.push(this._map(item));
        return this._items.length;
    }
    _addOneOrMore(items, options) {
        if (Array.isArray(items)) {
            if ((0, common_1.isDefined)(options) && (0, common_1.isDefined)(options.index)) {
                items = items.reverse();
            }
            for (const item of items) {
                this._addOne(item, options);
            }
        }
        else {
            this._addOne(items, options);
        }
        return this.length;
    }
    toArray() {
        return [...this._items];
    }
}
exports.Collection = Collection;
