// @fleet-sdk/core/dist/cjs/models/ergoBox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErgoBox = void 0;
const common_1 = require("@fleet-sdk/common");
const blake2b_1 = require("@noble/hashes/blake2b");
const utils_1 = require("@noble/hashes/utils");
const boxSerializer_1 = require("../serializer/sigma/boxSerializer");
class ErgoBox {
    boxId;
    value;
    ergoTree;
    creationHeight;
    assets;
    additionalRegisters;
    transactionId;
    index;
    constructor(box) {
        this.boxId = box.boxId;
        this.ergoTree = box.ergoTree;
        this.creationHeight = box.creationHeight;
        this.value = (0, common_1.ensureBigInt)(box.value);
        this.assets = box.assets.map((asset) => ({
            tokenId: asset.tokenId,
            amount: (0, common_1.ensureBigInt)(asset.amount)
        }));
        this.additionalRegisters = box.additionalRegisters;
        this.transactionId = box.transactionId;
        this.index = box.index;
    }
    isValid() {
        return ErgoBox.validate(this);
    }
    static validate(box) {
        const bytes = (0, boxSerializer_1.serializeBox)(box);
        const hash = (0, utils_1.bytesToHex)((0, blake2b_1.blake2b)(bytes, { dkLen: 32 }));
        return box.boxId === hash;
    }
}
exports.ErgoBox = ErgoBox;
