// @fleet-sdk/core/dist/cjs/models/ergoUnsignedInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErgoUnsignedInput = void 0;
const ergoBox_1 = require("./ergoBox");
class ErgoUnsignedInput extends ergoBox_1.ErgoBox {
    _extension;
    get extension() {
        return this._extension;
    }
    constructor(box) {
        super(box);
    }
    setContextVars(extension) {
        this._extension = extension;
        return this;
    }
    toUnsignedInputObject(type) {
        return {
            ...this.toObject(type),
            extension: this._extension || {}
        };
    }
    toObject(type) {
        if (type === "EIP-12") {
            return {
                boxId: this.boxId,
                value: this.value.toString(),
                ergoTree: this.ergoTree,
                creationHeight: this.creationHeight,
                assets: this.assets.map((asset) => ({
                    tokenId: asset.tokenId,
                    amount: asset.amount.toString()
                })),
                additionalRegisters: this.additionalRegisters,
                transactionId: this.transactionId,
                index: this.index
            };
        }
        else {
            return {
                boxId: this.boxId
            };
        }
    }
}
exports.ErgoUnsignedInput = ErgoUnsignedInput;
