// @fleet-sdk/core/dist/cjs/builder/selector/boxSelector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BoxSelector = void 0;
const common_1 = require("@fleet-sdk/common");
const common_2 = require("@fleet-sdk/common");
const duplicateInputSelectionError_1 = require("../../errors/duplicateInputSelectionError");
const insufficientInputs_1 = require("../../errors/insufficientInputs");
const accumulativeSelectionStrategy_1 = require("./strategies/accumulativeSelectionStrategy");
const customSelectionStrategy_1 = require("./strategies/customSelectionStrategy");
class BoxSelector {
    _inputs;
    _strategy;
    _ensureFilterPredicate;
    _inputsSortSelector;
    _inputsSortDir;
    _ensureInclusionBoxIds;
    constructor(inputs) {
        this._inputs = inputs;
    }
    defineStrategy(strategy) {
        if (this._isISelectionStrategyImplementation(strategy)) {
            this._strategy = strategy;
        }
        else {
            this._strategy = new customSelectionStrategy_1.CustomSelectionStrategy(strategy);
        }
        return this;
    }
    select(target) {
        if (!this._strategy) {
            this._strategy = new accumulativeSelectionStrategy_1.AccumulativeSelectionStrategy();
        }
        const remaining = this._deepCloneTarget(target);
        let unselected = [...this._inputs];
        let selected = [];
        const predicate = this._ensureFilterPredicate;
        const inclusion = this._ensureInclusionBoxIds;
        if ((0, common_2.isDefined)(predicate)) {
            if ((0, common_2.isDefined)(inclusion)) {
                selected = unselected.filter((box) => predicate(box) || inclusion.has(box.boxId));
            }
            else {
                selected = unselected.filter(predicate);
            }
        }
        else if ((0, common_2.isDefined)(inclusion)) {
            selected = unselected.filter((box) => inclusion.has(box.boxId));
        }
        if ((0, common_2.isDefined)(selected)) {
            unselected = unselected.filter((box) => !selected.some((sel) => sel.boxId === box.boxId));
            if ((0, common_2.isDefined)(remaining.nanoErgs)) {
                remaining.nanoErgs -= (0, common_2.sumBy)(selected, (input) => input.value);
            }
            if ((0, common_2.isDefined)(remaining.tokens) && selected.some((input) => !(0, common_2.isEmpty)(input.assets))) {
                for (const tokenTarget of remaining.tokens) {
                    if (tokenTarget.amount) {
                        tokenTarget.amount -= (0, common_2.utxoSum)(selected, tokenTarget.tokenId);
                    }
                }
            }
        }
        unselected = this._sort(unselected);
        selected = selected.concat(this._strategy.select(unselected, remaining));
        if ((0, common_2.hasDuplicatesBy)(selected, (item) => item.boxId)) {
            throw new duplicateInputSelectionError_1.DuplicateInputSelectionError();
        }
        const unreached = this._getUnreachedTargets(selected, target);
        if (unreached.nanoErgs || (0, common_2.some)(unreached.tokens)) {
            throw new insufficientInputs_1.InsufficientInputs(unreached);
        }
        return selected;
    }
    _deepCloneTarget(target) {
        return {
            nanoErgs: target.nanoErgs,
            tokens: (0, common_2.isDefined)(target.tokens)
                ? target.tokens.map((t) => ({ tokenId: t.tokenId, amount: t.amount }))
                : undefined
        };
    }
    _getUnreachedTargets(inputs, target) {
        const unreached = { nanoErgs: undefined, tokens: undefined };
        const selectedNanoergs = (0, common_2.sumBy)(inputs, (input) => input.value);
        if (target.nanoErgs && target.nanoErgs > selectedNanoergs) {
            unreached.nanoErgs = target.nanoErgs - selectedNanoergs;
        }
        if ((0, common_2.isEmpty)(target.tokens)) {
            return unreached;
        }
        for (const tokenTarget of target.tokens) {
            const totalSelected = (0, common_2.utxoSum)(inputs, tokenTarget.tokenId);
            if ((0, common_2.isDefined)(tokenTarget.amount) && tokenTarget.amount > totalSelected) {
                if (tokenTarget.tokenId === (0, common_1.first)(inputs).boxId) {
                    continue;
                }
                if ((0, common_1.isUndefined)(unreached.tokens)) {
                    unreached.tokens = [];
                }
                unreached.tokens.push({
                    tokenId: tokenTarget.tokenId,
                    amount: tokenTarget.amount - totalSelected
                });
            }
        }
        return unreached;
    }
    _sort(inputs) {
        if (!this._inputsSortSelector) {
            return (0, common_2.orderBy)(inputs, (input) => input.creationHeight, "asc");
        }
        return (0, common_2.orderBy)(inputs, this._inputsSortSelector, this._inputsSortDir || "asc");
    }
    ensureInclusion(predicateOrBoxIds) {
        if (typeof predicateOrBoxIds === "function") {
            this._ensureFilterPredicate = predicateOrBoxIds;
        }
        else {
            if ((0, common_1.isUndefined)(this._ensureInclusionBoxIds)) {
                this._ensureInclusionBoxIds = new Set();
            }
            if (Array.isArray(predicateOrBoxIds)) {
                for (const boxId of predicateOrBoxIds) {
                    this._ensureInclusionBoxIds.add(boxId);
                }
            }
            else {
                this._ensureInclusionBoxIds.add(predicateOrBoxIds);
            }
        }
        return this;
    }
    orderBy(selector, direction) {
        this._inputsSortSelector = selector;
        this._inputsSortDir = direction;
        return this;
    }
    _isISelectionStrategyImplementation(obj) {
        if (obj.select) {
            return true;
        }
        return false;
    }
    static buildTargetFrom(boxes) {
        const tokens = {};
        let nanoErgs = common_2._0n;
        for (const box of boxes) {
            nanoErgs += (0, common_2.ensureBigInt)(box.value);
            for (const token of box.assets) {
                tokens[token.tokenId] = (tokens[token.tokenId] || common_2._0n) + (0, common_2.ensureBigInt)(token.amount);
            }
        }
        return {
            nanoErgs,
            tokens: Object.keys(tokens).map((tokenId) => ({ tokenId, amount: tokens[tokenId] }))
        };
    }
}
exports.BoxSelector = BoxSelector;
